File: Benchmarks/DelannoyNumbers/Julia/ParallelWithCutoff.jl
using Base.Threads
using Statistics
using ThreadPinning 

pinthreads(:cores)

function delannoy_seq(m::Int, n::Int)
    if m == 0 || n == 0
        return 1
    else
        return delannoy_seq(m - 1, n) + delannoy_seq(m - 1, n - 1) + delannoy_seq(m, n - 1)
    end      
end


function delannoy(m::Int, n::Int)
    if m == 0 || n == 0
        return 1    
    elseif m < 9 || n < 9
        return delannoy_seq(m, n)
    end
    
    task1 = @spawn delannoy(m - 1, n)

    task2 = @spawn delannoy(m - 1, n - 1)

    task3 = @spawn delannoy(m, n - 1)

    return  fetch(task1) + fetch(task2) + fetch(task3)
end


if length(ARGS) != 1
    println(stderr, "Usage: julia DelannoySeq.jl <N>")
    exit(1)
end

str = ARGS[1]

const n = parse(Int, str)
const m = n

const N_RUNS = 20
times = Float64[]

delannoy(m, n)
delannoy(m, n)
delannoy(m, n)

for i in 1:N_RUNS
    t = @elapsed result = delannoy(m, n)
    push!(times, t)
    println(result)
end

println("--- Times ---")
println(times)

println("\n--- Statistics ---")
println("Average:     $(mean(times)) s")

File: Benchmarks/DelannoyNumbers/C/ParallelWithCutoff.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

long delannoy_seq(int m, int n) {
    if (m == 0 || n == 0) {
        return 1;
    } else {
        return delannoy_seq(m - 1, n) + delannoy_seq(m - 1, n - 1) + delannoy_seq(m, n - 1);
    }
}

long delannoy_par(int m, int n) {
    long result1, result2, result3;

    if (m == 0 || n == 0) {
        return 1;
    } else if (m < 9 || n < 9) {
        return delannoy_seq(m, n);
    }

    #pragma omp task shared(result1) 
    {
        result1 = delannoy_par(m - 1, n);
    }
    #pragma omp task shared(result2) 
    {
        result2 = delannoy_par(m - 1, n - 1);
    }

    #pragma omp task shared(result3)
    {
        result3 = delannoy_par(m, n - 1);
    }
    

    #pragma omp taskwait
    return result1 + result2 + result3;
}


int main(int argc, char * argv[]) {

    if (argc != 2) {
        fprintf(stderr, "Usage: [n]\n");
		return EXIT_FAILURE;
    }

    int n = atoi(argv[1]);

    int n_runs = 20;
    double *times = malloc(n_runs * sizeof(double));

    for (int run = 0; run < n_runs; run++) {
        double start = omp_get_wtime();

        long result;
        #pragma omp parallel
        {
            #pragma omp single
            {
                result = delannoy_par(n, n);
            }
        }
        double end = omp_get_wtime();
        times[run] = end - start;
        printf("%ld\n", result);
    }

    printf("--- Times ---\n");
    double sum = 0.0;
    printf("[");

    for (int run = 0; run < n_runs; run++) {
        if (run > 0) {
            printf(", ");
        }
        printf("%f", times[run]);
        sum += times[run];
    }

    printf("]\n");

    printf("\n--- Statistics ---\n");
    printf("Average: %f\n", sum / n_runs);

    free(times);

    return EXIT_SUCCESS;
}